src/functional/fencil_processors/builders/cpp/build.py

```diff
@@ -0,0 +1,207 @@
+        """
+        project({{project_name}})
+        cmake_minimum_required(VERSION 3.20.0)
+
+        # Languages
+        enable_language(CXX)
+
+        # Paths
+        list(APPEND CMAKE_MODULE_PATH ${CMAKE_BINARY_DIR})
+        list(APPEND CMAKE_PREFIX_PATH ${CMAKE_BINARY_DIR})
+
+        set(CMAKE_ARCHIVE_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+        set(CMAKE_LIBRARY_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+        set(CMAKE_RUNTIME_OUTPUT_DIRECTORY ${CMAKE_BINARY_DIR}/bin)
+
+        # Find dependencies
+        include(FetchContent)
+
+        {{"\\n".join(find_deps)}}
+
+        # Targets
+        add_library({{project_name}} MODULE)
+
+        target_compile_features({{project_name}} PRIVATE cxx_std_17)
+        set_target_properties({{project_name}} PROPERTIES PREFIX "" SUFFIX ".{{bin_output_suffix}}")
+
+        target_sources({{project_name}}
+            PRIVATE
+                {{"\\n".join(source_names)}}
+        )
+
+        # Link dependencies
+        {{"\\n".join(link_deps)}}
+        """
+    )
+
+    def visit_FindDependency(self, dep: FindDependency):
+        match dep.name:
+            case "pybind11":
+                import pybind11
+
+                return f"find_package(pybind11 CONFIG REQUIRED PATHS {pybind11.get_cmake_dir()})"
+            case "gridtools":
+                return textwrap.dedent(
+                    """\
+                    FetchContent_Declare(GridTools
+                        GIT_REPOSITORY https://github.com/GridTools/gridtools.git
+                        GIT_TAG        master
+                    )
+                    FetchContent_MakeAvailable(GridTools)\
+                    """
+                )
+            case _:
+                raise ValueError("Library {name} is not supported".format(name=dep.name))
+
+    def visit_LinkDependency(self, dep: LinkDependency):
+        match dep.name:
+            case "pybind11":
+                lib_name = "pybind11::module"
+            case "gridtools":
+                lib_name = "GridTools::fn_naive"
+            case _:
+                raise ValueError("Library {name} is not supported".format(name=dep.name))
+        return "target_link_libraries({target} PUBLIC {lib})".format(
+            target=dep.target, lib=lib_name
+        )
+
+
+def _render_cmakelists(
+    project_name: str,
+    dependencies: Sequence[source_modules.LibraryDependency],
+    source_names: Sequence[str],
+) -> str:
+    cmakelists_file = CMakeListsFile(
+        project_name=project_name,
+        find_deps=[FindDependency(name=d.name, version=d.version) for d in dependencies],
+        link_deps=[LinkDependency(name=d.name, target=project_name) for d in dependencies],
+        source_names=source_names,
+        bin_output_suffix=_get_python_module_suffix(),
+    )
+    return CMakeListsGenerator.apply(cmakelists_file)
+
+
+def _get_python_module_suffix():
+    return importlib.machinery.EXTENSION_SUFFIXES[0][1:]
+
+
+class CMakeProject:
```

Comment:
My understanding may be limited, but is `dataclass` not basically a `struct`? `CMakeProject` has a rather rich functionality, non-trivial construction and mutable state, so I didn't find `dataclass` fitting here.

Reply:
Basically, dataclass just generates some data members related magic methods for you, like `__init__`. A frozen dataclass additionally ensures that the data members can not be changed after construction. Nothing stops a `dataclass` from having methods or any other class functionality.

Reply:
Yes, exactly, the `@dataclass` decorator adds the missing boilerplate methods to your class, but doesn't overwrite the methods you define manually or prevents you to add more methods. In this case it will basically add proper `__eq__` and `__repr__` methods, since you define a custom `__init__`.

Reply:
Will leave it as a non-dataclass due to its current form.